### PR TITLE
Added model training script

### LINKED ISSUE #9: Fix validation dataset transform bug
**Acceptance Criteria:**

- [x] Augmentations/transformations are applied *only* to the training dataset.
- [x] The validation pipeline works correctly without being affected by training transformations.
- [x] Validation metrics are reliable and calculated properly.
- [x] The full code execution completes successfully without errors.

### LINKED ISSUE #10: Implement Dice + Focal Loss or Tversky Loss for segmentation
Improve the loss function used for UNet++ training.
Focal Loss focuses training on difficult pixels (e.g. boundaries between optic disc and optic cup).  
Tversky Loss allows separate weighting of false positives and false negatives, which is important in the clinical context where missing glaucoma cases (false negatives) is more critical than false alarms.
- Implement Dice Loss
- Implement Focal Loss or Tversky Loss
- Integrate the loss functions into the training pipeline

### LINKED ISSUE #11: Implement Mixed Precision Training (AMP)
Mixed precision allows increasing the batch size (e.g. from 8 to 16–24), which improves training stability. Larger batches stabilize Batch Normalization layers in the encoder and can lead to better Dice scores.
- Integrate PyTorch AMP into the training loop
- Validate numerical stability

### PR DESCRIPTION
**Pull request by MateuszDobry**
Added Unet++ Model

Closes #9 
Closes #10 
Closes #11 

## Release Notes

* **New Features**
  * Added U-Net++ training for glaucoma disc and cup segmentation from fundus images
  * Implemented automated image augmentation pipeline with multiple transformation options
  * Support for multiple configurable loss functions including weighted Dice-Focal combinations
  * Early stopping and best model checkpoint management for training efficiency
  * Cross-validation support for robust model evaluation across multiple folds

### CODE REVIEW

File: ai/unet/Model Training.py [hunk 1]
```diff
@@ -0,0 +1,528 @@
...
+        return len(self.image_files)
+
+    def __getitem__(self, idx):
+        img_path = self.image_files[idx]
+        image    = cv2.cvtColor(cv2.imread(str(img_path)), cv2.COLOR_BGR2RGB)
+
+        mask_path = self.masks_dir / f"{img_path.stem}.png"
+        if not mask_path.exists():
+            raise FileNotFoundError(f"Missing mask: {mask_path}")
+        mask = cv2.imread(str(mask_path), cv2.IMREAD_GRAYSCALE)
```
[the remark this answers is not in the document]
Comment (kequel) [reply ai/unet/Model Training.py:99]: I think validation here would indeed be safer

File: ai/unet/Model Training.py [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+        m = cv2.morphologyEx(mask.astype(np.uint8), cv2.MORPH_CLOSE, kernel)
+        n, labels, stats, _ = cv2.connectedComponentsWithStats(m)
+        if n > 1:
+            m = (labels == (1 + np.argmax(stats[1:, cv2.CC_STAT_AREA]))).astype(np.uint8)
+        return m.astype(np.float32)
+
+    disc = clean(pred_disc_np)
+    cup  = clean(pred_cup_np) * disc   # cup only where disc is present
+    return disc, cup
+
```
[the remark this answers is not in the document]
Comment (kequel) [reply ai/unet/Model Training.py:294]: I do not really get it, @MateuszDobry is this compatible with Antek's pipeline or not?

**Review by Amila321 — APPROVED** [review]

**Review by martynkaqhe — APPROVED** [review]
